fido2/hid/freebsd.py

```diff
@@ -28,20 +30,105 @@
 logger = logging.getLogger(__name__)
 
 
+devdir = "/dev/"
+
+device_re = re.compile("^uhid([0-9]+)$")
+vendor_re = re.compile("vendor=(0x[0-9a-fA-F]+)")
+product_re = re.compile("product=(0x[0-9a-fA-F]+)")
+sernum_re = re.compile('sernum="([^"]+)')
+
+libc = ctypes.CDLL(find_library("c"))
+
+USB_GET_REPORT_DESC = 0xC0205515
+
+
+class usb_gen_descriptor(ctypes.Structure):
+    _fields_ = [
+        (
+            "ugd_data",
+            ctypes.c_void_p,
+        ),  # TODO: check what COMPAT_32BIT in C header means
+        ("ugd_lang_id", ctypes.c_uint16),
+        ("ugd_maxlen", ctypes.c_uint16),
+        ("ugd_actlen", ctypes.c_uint16),
+        ("ugd_offset", ctypes.c_uint16),
+        ("ugd_config_index", ctypes.c_uint8),
+        ("ugd_string_index", ctypes.c_uint8),
+        ("ugd_iface_index", ctypes.c_uint8),
+        ("ugd_altif_index", ctypes.c_uint8),
+        ("ugd_endpt_index", ctypes.c_uint8),
+        ("ugd_report_type", ctypes.c_uint8),
+        ("reserved", ctypes.c_uint8 * 8),
+    ]
+
+
 def open_connection(descriptor):
     return FileCtapHidConnection(descriptor)
 
 
+def _get_report_data(fd, report_type):
+    data = ctypes.create_string_buffer(b"\000" * 4096)
+    desc = usb_gen_descriptor(
+        ugd_data=ctypes.addressof(data),
+        ugd_maxlen=ctypes.sizeof(data),
+        report_type=report_type,
+    )
+    ret = libc.ioctl(fd, USB_GET_REPORT_DESC, ctypes.byref(desc))
+    if ret != 0:
+        raise ValueError("ioctl failed")
+    return data.raw[: desc.ugd_actlen]
+
+
 def _read_descriptor(vid, pid, name, serial, path):
     fd = os.open(path, os.O_RDONLY)
-    data = uhid_freebsd.get_report_data(fd, 3)
+    data = _get_report_data(fd, 3)
     os.close(fd)
     max_in_size, max_out_size = parse_report_descriptor(data)
     return HidDescriptor(path, vid, pid, max_in_size, max_out_size, name, serial)
 
 
+def _enumerate():
+    for uhid in os.listdir(devdir):
```

Comment:
I'd replace this with `glob`, similar to https://github.com/Yubico/python-fido2/blob/master/fido2/hid/linux.py#L81
That doesn't cause failure if for any reason `/dev` isn't readable and allows us to drop `device_re`.

Reply:
I have changed the `_enumerate()` accordingly.